### PR TITLE
Don't use explicit mangles for builtins

### PR DESCRIPTION
**Pull request by thewilsonator**
This also fixes `std/math.d(4712): Error: `ldexp` cannot be interpreted at compile time, because it has no available source code` for https://github.com/dlang/phobos/pull/7463

### CODE REVIEW

**Comment by thewilsonator** [comment]
Gah, mixed integer/float `pow` needs to be fixed.

**Review by ibuclaw — APPROVED** [review]
Maybe it's time to consider pragma(builtin) for marking intrinsics?

File: src/dmd/builtin.d [hunk 1]
```diff
@@ -40,20 +75,91 @@ private:
...
+    if (id3 == Id._pow)
+    {
+        if ((*fd.parameters)[0].type.isfloating() &&
+            (*fd.parameters)[1].type.isfloating())
+            return BUILTIN.pow;
+        return BUILTIN.unimp;
+    }
 
-builtin_fp builtin_lookup(const(char)* mangle)
-{
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply src/dmd/builtin.d:53]: ?

File: src/dmd/declaration.h [hunk 2]
```diff
@@ -442,14 +442,45 @@ enum ILS
 
 enum BUILTIN
 {
-    BUILTINunknown = -1,        // not known if this is a builtin
-    BUILTINno,                  // this is not a builtin
-    BUILTINyes                  // this is a builtin
+    BUILTINunknown = -1,    /// not known if this is a builtin
```
Comment (WalterBright) [inline src/dmd/declaration.h:445]: Prefer to not use mixed sign for enum, as that interferes with efficient switch dispatch and some other optimizations.

File: src/dmd/declaration.h [hunk 3]
```diff
@@ -442,14 +442,45 @@ enum ILS
 
 enum BUILTIN
```
Comment (WalterBright) [inline src/dmd/declaration.h:443]: ` : ubyte`

File: src/dmd/builtin.d [hunk 4]
```diff
@@ -25,8 +25,43 @@ import dmd.mtype;
...
+
+/**************************************
+ * Evaluate builtin function.
+ * Return result; NULL if cannot evaluate it.
+ */
+public extern (C++) Expression eval_builtin(Loc loc, FuncDeclaration fd, Expressions* arguments)
+{
+    if (fd.builtin == BUILTIN.unknown)
+        fd.builtin = determine_builtin(fd);
+    if (fd.builtin == BUILTIN.unimp)
```
Comment (WalterBright) [inline src/dmd/builtin.d:52]: This test should be first as it is most popular.

File: src/dmd/builtin.d [hunk 5]
```diff
@@ -25,8 +25,43 @@ import dmd.mtype;
...
+    return fd.builtin;
+}
+
+/**************************************
+ * Evaluate builtin function.
+ * Return result; NULL if cannot evaluate it.
+ */
+public extern (C++) Expression eval_builtin(Loc loc, FuncDeclaration fd, Expressions* arguments)
+{
+    if (fd.builtin == BUILTIN.unknown)
```
Comment (WalterBright) [inline src/dmd/builtin.d:50]: This is duplicating the test in isBuiltin.

File: src/dmd/builtin.d [hunk 6]
```diff
@@ -25,8 +25,43 @@ import dmd.mtype;
...
+        fd.builtin = determine_builtin(fd);
+    if (fd.builtin == BUILTIN.unimp)
+        return null;
+
+    switch (fd.builtin) {
```
Comment (WalterBright) [inline src/dmd/builtin.d:55]: Don't use egyptian bracing style

File: src/dmd/builtin.d [hunk 7]
```diff
@@ -25,8 +25,43 @@ import dmd.mtype;
...
+        foreach(e; __traits(allMembers, BUILTIN))
+        {
+            static if (e == "unknown"){ case BUILTIN.unknown: assert(false); }
```
Comment (WalterBright) [inline src/dmd/builtin.d:58]: format as:
```
static if (cond)
   ...
else
   ...
```

File: src/dmd/builtin.d (hunk 4 above)
Comment (thewilsonator) [reply src/dmd/builtin.d:52]: Can't do that. We need to test if it has been tested before. Note that `determine_builtin` may well return `BUILTIN.unimp`

File: src/dmd/builtin.d [hunk 8]
```diff
@@ -40,20 +75,91 @@ private:
...
 
-__gshared StringTable!builtin_fp builtins;
-
-void add_builtin(const(char)[] mangle, builtin_fp fp)
-{
-    builtins.insert(mangle, fp);
-}
+BUILTIN determine_builtin(FuncDeclaration fd)
+{
+    fd = fd.toAliasFunc();
```
Comment (WalterBright) [inline src/dmd/builtin.d:81]: Create new variable instead of assigning a new value to the parameter.

File: src/dmd/builtin.d [hunk 9]
```diff
@@ -40,20 +75,91 @@ private:
...
+    const id3 = fd.ident;
+    auto m = fd.getModule();
+    if (!m || !m.md)
+        return BUILTIN.unimp;
+    const md = m.md;
+    const id2 = md.id;
+
+    if (!md.packages)
+        return BUILTIN.unimp;
+    if (md.packages.dim != 1)
```
Comment (WalterBright) [inline src/dmd/builtin.d:93]: use `.length` instead of `.dim`

File: src/dmd/builtin.d [hunk 10]
```diff
@@ -40,20 +75,91 @@ private:
...
+        return BUILTIN.unimp;
+    // Look for core.math, core.bitop and std.math
+    const id1 = (*md.packages)[0];
+    if (id1 != Id.core && id1 != Id.std)
+        return BUILTIN.unimp;
+    if ((id2 == Id.bitop && id1 != Id.core))
+        return BUILTIN.unimp;
+    else if (id2 != Id.math)
+        return BUILTIN.unimp;
+    if (id2 == Id.bitop)
```
Comment (WalterBright) [inline src/dmd/builtin.d:103]: Multiple instances of redundant comparisons. `bitop` and `core`

File: src/dmd/builtin.d [hunk 11]
```diff
@@ -40,20 +75,91 @@ private:
...
+{
+    fd = fd.toAliasFunc();
+    if (fd.isDeprecated())
+        return BUILTIN.unimp;
+    const id3 = fd.ident;
+    auto m = fd.getModule();
+    if (!m || !m.md)
+        return BUILTIN.unimp;
+    const md = m.md;
+    const id2 = md.id;
```
Comment (WalterBright) [inline src/dmd/builtin.d:89]: early exit here if id2 is not bitop or math.

File: src/dmd/builtin.d [hunk 12]
```diff
@@ -40,20 +75,91 @@ private:
...
-void add_builtin(const(char)[] mangle, builtin_fp fp)
-{
-    builtins.insert(mangle, fp);
-}
+BUILTIN determine_builtin(FuncDeclaration fd)
+{
+    fd = fd.toAliasFunc();
+    if (fd.isDeprecated())
+        return BUILTIN.unimp;
+    const id3 = fd.ident;
```
Comment (WalterBright) [inline src/dmd/builtin.d:84]: don't set id3 until needed

**Comment by WalterBright** [comment]
Looks like a sensible replacement for the existing code.

File: src/dmd/declaration.h (hunk 3 above)
Comment (thewilsonator) [reply src/dmd/declaration.h:443]: Thats a header file and we need C++98 compatibility.

File: src/dmd/declaration.h (hunk 2 above)
Comment (thewilsonator) [reply src/dmd/declaration.h:445]: That was there before and I'd rather not make unnecessary changes. Feel free to follow up on that though.

**Pull request merged**